ocaml/xapi/xapi_pool.ml

```diff
@@ -591,9 +591,18 @@ let create_or_get_pif_on_master __context rpc session_id (pif_ref, pif) : API.re
   let my_host = Db.Host.get_record ~__context ~self:my_host_ref in
   let new_host_ref = create_or_get_host_on_master __context rpc session_id (my_host_ref, my_host) in
 
-  let my_network_ref = pif.API.pIF_network in
-  let my_network = Db.Network.get_record ~__context ~self:my_network_ref in
-  let new_network_ref = create_or_get_network_on_master __context rpc session_id (my_network_ref, my_network) in
+  let new_network_ref =
+    if pif.API.pIF_VLAN <> -1L then begin
+      (* Get the remote management network for management VLAN PIF *)
+      let remote_mgmt_pif = Client.Host.get_management_interface rpc session_id (get_master ~rpc ~session_id) in
```

Comment:
This would mean that `create_or_get_pif_on_master` only works for a VLAN PIF if it is a management PIF. It can easily be made more general by finding a PIF on the master with the same `VLAN` and `device`, and use its network. If none is found, then just call `create_or_get_network_on_master` (though this shouldn't occur at the moment, because the pre-join checks prevent it).